cgfs.py

```diff
@@ -365,46 +502,46 @@ def mkdir(self, path, mode):
         if dname in parent.children:  # pragma: no cover
             raise FuseOSError(EEXIST)
 
-        submission = self.get_submission(path)
-        query_path = submission.tld + '/' + '/'.join(parts[3:]) + '/'
-        ddata = cgapi.create_file(submission.id, query_path)
+        if self.fixed:
+            parent.insert(TempDirectory({}, name=dname, writable=True))
+        else:
+            submission = self.get_submission(path)
+            query_path = submission.tld + '/' + '/'.join(parts[3:]) + '/'
+            ddata = cgapi.create_file(submission.id, query_path)
 
-        parent.insert(Directory(ddata, name=dname, writable=True))
+            parent.insert(Directory(ddata, name=dname, writable=True))
 
     def open(self, path, flags):
         parts = self.split_path(path)
         parent = self.get_dir(parts[:-1])
 
-        try:
-            file = self.get_file(parts[-1], start=parent, expect_type=File)
-        except FuseOSError as e:
-            if e.errno is not ENOENT:
-                raise e
-            if not flags & O_CREAT:
-                raise FuseOSError(ENOENT)
-            return self.create(path, 0o770)
-        else:
-            # This should be handled by FUSE.
-            if flags & (O_CREAT & O_EXCL):  # pragma: no cover
-                raise FuseOSError(EEXIST)
+        file = self.get_file(parts[-1], start=parent, expect_type=File)
 
-        if file.data is None:
-            file.open(cgapi.get_file(file.id))
+        if isinstance(file, TempFile):
+            file.open()
 
-        if flags & O_TRUNC:
+        # This is handled by fuse [0] but it can be disabled so it is better to
+        # be save than sorry as it can be enabled.
+        # [0] https://sourceforge.net/p/fuse/mailman/message/29515577/
+        if flags & O_TRUNC:  # pragma: no cover
             file.truncate(0)
 
         self.fd += 1
+        self._open_files[self.fd] = file
         return self.fd
 
     def read(self, path, size, offset, fh):
-        file = self.get_file(path, expect_type=File)
+        file = self._open_files[fh]
+
+        if isinstance(file, TempFile):
+            return file.read(offset, size)
+
         return file.data[offset:offset + size]
```

Comment:
Shouldn't we make a `read` method on the `File` class instead of special-casing the `TempFile`?